ADAL/src/cache/ios/ADKeychainTokenCache.m

```diff
@@ -245,30 +259,138 @@ - (ADTokenCacheItem*)itemFromKeyhainAttributes:(NSDictionary*)attrs
+    [self logItem:item message:@"Item being tombstoned"];
+    
+    ADTokenCacheItem* tombstoneItem = [item copy];
+    SAFE_ARC_AUTORELEASE(tombstoneItem);
+    [self setCacheItemAsTombstone:tombstoneItem];
     
+    //update tombstone in cache
+    BOOL updateStatus = [self addOrUpdateItem:tombstoneItem error:error];
+    
+    return updateStatus;
+    
+}
+
+//Interal function: delete an item from keychain;
+- (OSStatus)deleteItem:(nonnull ADTokenCacheItem *)item
+             error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    RETURN_NO_ON_NIL_ARGUMENT(item);
     ADTokenCacheKey* key = [item extractKey:error];
     if (!key)
     {
         return NO;
     }
-    
-    
     NSMutableDictionary* query = [self queryDictionaryForKey:key
                                                       userId:item.userInformation.userId
                                                   additional:nil];
-    OSStatus status = SecItemDelete((CFDictionaryRef)query);
-    return [ADKeychainTokenCache checkStatus:status details:@"Failed to remove item from keychain" error:error];
+    return SecItemDelete((CFDictionaryRef)query);
+}
+
+-(void)setCacheItemAsTombstone:(ADTokenCacheItem*)item
+{
+    if (item)
+    {
+        //avoid bundleId being nil, as it will be stored in a NSMutableDictionary
+        NSString* bundleId = [[NSBundle mainBundle] bundleIdentifier] ? [[NSBundle mainBundle] bundleIdentifier] : @"";
+        
+        if (item.tombstone)
+        {
+            [item.tombstone addEntriesFromDictionary:@{@"bundleId" : bundleId}];
+        }
+        else
+        {
+            [item setTombstone:[NSMutableDictionary dictionaryWithDictionary:@{@"bundleId" : bundleId}]];
+        }
+        
+        //wipe out the refresh token
+        [item setRefreshToken:@"<tombstone>"];
+    }
+}
+
+-(NSMutableArray*)filterOutTombstones:(NSArray*) items
+{
+    if (!items)
+    {
+        return nil;
+    }
+    
+    NSMutableArray* itemsKept = [NSMutableArray new];
+    for (ADTokenCacheItem* item in items)
+    {
+        if (![item tombstone])
+        {
+            [itemsKept addObject:item];
+        }
+    }
+    SAFE_ARC_AUTORELEASE(itemsKept);
+    return itemsKept;
+}
+
+- (BOOL)removeAllForClientId:(NSString * __nonnull)clientId
+                       error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    AD_LOG_VERBOSE_F(@"Keychain token cache store", nil, @"Removing items for client <%@>", clientId);
+    
+    BOOL deleteSuccessful = YES;
+    NSArray* items = [self allItems:nil];
+    for (ADTokenCacheItem * item in items)
+    {
+        if ([clientId isEqualToString:[item clientId] ])
+        {
+            [self removeItem:item error:error];
+            if (*error)
+            {
+                deleteSuccessful = NO;
+            }
+        }
+    }
+    return deleteSuccessful;
+}
+
+- (BOOL)removeAllForIdentifier:(NSString * __nonnull)userId
```

Comment:
Change this to "ForUserId:" we use "Identifier" when we take in a ADUserIdentifier.